registry.go

```diff
@@ -17,152 +17,247 @@ package schemaregistry
 )
 
 // SchemaRegistry is a schema registry that stores schemas in memory.
 type SchemaRegistry struct {
-	schemas            []sr.SubjectSchema
-	fingerprintIDCache map[uint64]int
-	idSequence         int
+	db database.DB
+
+	schemaStore        *schemaStore
+	subjectSchemaStore *subjectSchemaStore
+	idSequenceStore    *sequenceStore
+
+	fingerprintIDCache   map[uint64]int
+	idSubjectSchemaCache map[int][]sr.SubjectSchema // subject schema doesn't contain actual schema
+	subjectVersionCache  map[string][]int
+
+	idSequence int
 
 	m sync.Mutex
 }
 
-func NewSchemaRegistry() *SchemaRegistry {
-	return &SchemaRegistry{
-		schemas:            make([]sr.SubjectSchema, 0),
-		fingerprintIDCache: make(map[uint64]int),
+func NewSchemaRegistry(db database.DB) (*SchemaRegistry, error) {
+	r := &SchemaRegistry{
+		db: db,
+
+		schemaStore:        newSchemaStore(db),
+		subjectSchemaStore: newSubjectSchemaStore(db),
+		idSequenceStore:    newSequenceStore(db, "schemaid"),
+
+		fingerprintIDCache:   make(map[uint64]int),
+		subjectVersionCache:  make(map[string][]int),
+		idSubjectSchemaCache: make(map[int][]sr.SubjectSchema),
+	}
+
+	ctx := context.Background()
+
+	// load sequence
+	id, err := r.idSequenceStore.Get(ctx)
+	if err != nil && !errors.Is(err, database.ErrKeyNotExist) {
+		return nil, fmt.Errorf("failed to get schema id sequence from store: %w", err)
+	}
+	r.idSequence = id
+
+	// load schemas
+	sss, err := r.subjectSchemaStore.GetAll(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get all subject schemas from store: %w", err)
+	}
+	for _, ss := range sss {
+		fp := rabin.Bytes([]byte(ss.Schema.Schema))
+		r.fingerprintIDCache[fp] = ss.ID
+		r.subjectVersionCache[ss.Subject] = append(r.subjectVersionCache[ss.Subject], ss.Version)
+		r.idSubjectSchemaCache[ss.ID] = append(r.idSubjectSchemaCache[ss.ID], sr.SubjectSchema{
+			Subject: ss.Subject,
+			Version: ss.Version,
+			ID:      ss.ID,
+			Schema:  sr.Schema{}, // don't include schema in cache
+		})
 	}
+
+	return r, nil
 }
 
-func (r *SchemaRegistry) CreateSchema(ctx context.Context, subject string, schema sr.Schema) (sr.SubjectSchema, error) {
+//nolint:funlen // refactor at some point
+func (r *SchemaRegistry) CreateSchema(ctx context.Context, subject string, schema sr.Schema) (_ sr.SubjectSchema, err error) {
 	r.m.Lock()
 	defer r.m.Unlock()
 
-	fp := internal.Rabin([]byte(schema.Schema))
+	var rb rollback.R
+	defer rb.MustExecute()
+
+	fp := rabin.Bytes([]byte(schema.Schema))
 	id, ok := r.fingerprintIDCache[fp]
-	if ok {
-		// schema exists, see if subject matches
-		ss, err := r.findBySubjectID(subject, id)
-		if err == nil {
-			// schema exists for this subject, return it
-			return ss, nil
-		}
-		if !errors.Is(err, ErrSchemaNotFound) {
-			// unexpected error
-			return sr.SubjectSchema{}, err
-		}
-	} else {
+	if !ok {
 		// schema does not exist yet
 		id = r.nextID()
+		rb.AppendPure(r.revertID)
+	} else {
+		// check if a subject exists for this id
```

Comment:
Nit: might be useful to mention in the comment that we're looking for the latest version